Sources/NIO/SocketOptionChannel.swift

```diff
@@ -0,0 +1,147 @@
+    /// - returns: An `EventLoopFuture` that fires when the option has been set,
+    ///     or if an error has occurred.
+    func unsafeSetSocketOption<Value>(level: SocketOptionLevel, name: SocketOptionName, value: Value) -> EventLoopFuture<Void>
+
+    /// Obtain the value of the socket option for the given level and name.
+    ///
+    /// This function is not memory-safe: if you set the generic type parameter incorrectly,
+    /// this function will still execute, and this can cause you to incorrectly interpret memory
+    /// and thereby read uninitialized or invalid memory. If at all possible, please use one of
+    /// the safe functions defined by this protocol.
+    ///
+    /// - parameters:
+    ///     - level: The socket option level, e.g. `SOL_SOCKET` or `IPPROTO_IP`.
+    ///     - name: The name of the socket option, e.g. `SO_REUSEADDR`.
+    /// - returns: An `EventLoopFuture` containing the value of the socket option, or
+    ///     any error that occurred while retrieving the socket option.
+    func unsafeGetSocketOption<Value>(level: SocketOptionLevel, name: SocketOptionName) -> EventLoopFuture<Value>
+}
+
+
+// MARK:- Safe helper methods.
+// Hello code reader! All the methods in this extension are "safe" wrapper methods that define the correct
+// types for `setSocketOption` and `getSocketOption` and call those methods on behalf of the user. These
+// wrapper methods are memory safe. All of these methods are basically identical, and have been copy-pasted
+// around. As a result, if you change one, you should probably change them all.
+//
+// You are welcome to add more helper methods here, but each helper method you add must be tested.
+public extension SocketOptionChannel {
+    /// Sets the socket option SO_LINGER to `value`.
+    ///
+    /// - parameters:
+    ///     - value: The value to set SO_LINGER to.
+    /// - returns: An `EventLoopFuture` that fires when the option has been set,
+    ///     or if an error has occurred.
+    public func setSoLinger(_ value: linger) -> EventLoopFuture<Void> {
+        return self.unsafeSetSocketOption(level: SocketOptionLevel(SOL_SOCKET), name: SO_LINGER, value: value)
+    }
+
+    /// Gets the value of the socket option SO_LINGER.
+    ///
+    /// - returns: An `EventLoopFuture` containing the value of the socket option, or
+    ///     any error that occurred while retrieving the socket option.
+    public func getSoLinger() -> EventLoopFuture<linger> {
+        return self.unsafeGetSocketOption(level: SocketOptionLevel(SOL_SOCKET), name: SO_LINGER)
+    }
+}
+
+
+extension BaseSocketChannel: SocketOptionChannel {
+    public func unsafeSetSocketOption<Value>(level: SocketOptionLevel, name: SocketOptionName, value: Value) -> EventLoopFuture<Void> {
+        if eventLoop.inEventLoop {
+            let promise: EventLoopPromise<Void> = eventLoop.newPromise()
+            executeAndComplete(promise) { try setSocketOption0(level: level, name: name, value: value) }
```

Comment:
@normanmaurer/@lukasa what do you think about not having a new line after `{` ? I think we generally don't do it in NIO but I also don't mind it too much (but I have certainly made external contributors add extra newlines).

Reply:
Eh, I just added the newlines in.